## Group Assignment Week 8

In [1]:
import numpy as np
import math

#### Problem 1

In [2]:
# Define the data
year = np.arange(1994,2010,1)
year_norm = year - year.mean() 
ppm = np.array([356.8,358.2,360.3,361.8,364.0,365.7,366.7,368.2,370.5,372.2,374.9,376.7,378.7,381.0,382.9,384.7])

In [3]:
# Define the linear fitting function
def lineFit(x, y):
    xavg = x.mean()
    slope = (y * (x-xavg)).sum()/(x * (x-xavg)).sum()
    yint = y.mean() - slope*xavg
    return slope, yint, xavg

In [4]:
a, b, c =lineFit(year_norm,ppm)
print(f'a = {a}\nb = {b}')

a = 1.872205882352939
b = 370.20624999999995


<font size = 3 color = black face = "Times New Roman">The average increase of the concentration per year is 1.872205882352939 

#### Problem 2

<font size = 3 color = black face = "Times New Roman">We can think this problem as $f(x) = a_0f_0(x)+a_1f_1(x)$

In [5]:
# Define the data.
x = np.array([-0.5,-0.19,0.02,0.20,0.35,0.50])
y = np.array([-3.558,-2.874,-1.995,-1.040,-0.068,0.677])
f_0 = np.sin(math.pi * x / 2 )
f_1 = np.cos(math.pi * x / 2 )

#### Normal equation method.

In [6]:
# Defin two matrixs.
A = np.array([[(f_0 * f_0).sum(), (f_0 * f_1).sum()],[(f_0 * f_1).sum(), (f_1 * f_1).sum()]])
b_matrix = np.array([(f_0 * y).sum(), (f_1 * y).sum()])

In [7]:
a_matrix =  np.dot(np.linalg.inv(A),b)
print(f'a = {a_matrix[0]}\nb = {a_matrix[1]}')

a = [263.83782136 -28.43591508]
b = [-28.43591508  84.53518306]


#### QR method.

In [8]:
# Define two matrices.
A = np.zeros((6,2))
for i in range (6):
    A[i][0] = f_0[i]
    A[i][1] = f_1[i]
b_matrix = y.copy()

In [9]:
q,r = np.linalg.qr(A)
p = np.dot(q.T , b_matrix)
a_matrix = np.dot(np.linalg.inv(r),p)
print(f'a = {a_matrix[0]}\nb = {a_matrix[1]}')

a = 3.038490530950666
b = -2.0495596632192274


<font size = 3 color = black face = "Times New Roman">We can see that the results of these two methods are almost exactly the same, with only a small error, so these two methods can solve the problem well and get a good result.

#### Problem 3


#### Question 1

$f(x,y)= a + bx + cy$

<font size = 3 color = black face = "Times New Roman">Then the minimizing $S$ yields is:
$$
S = \sum[z_i - (a + bx_i+cy_i)]^2\\
\frac{\partial S}{\partial a}= - 2 \sum[z_i - (a + bx_i+cy_i)] = 0\\
\frac{\partial S}{\partial b}= - 2 \sum[z_i - (a + bx_i+cy_i)] x_i = 0\\
\frac{\partial S}{\partial c}= - 2 \sum[z_i - (a + bx_i+cy_i)] y_i= 0
$$
<font size = 3 color = black face = "Times New Roman">Hence we are easy to know that
$$
\begin{pmatrix}
n && \sum x_i && \sum y_i\\
\sum x_i && \sum x_i^2 && \sum x_i y_i\\
\sum y_i && \sum x_i y_i && \sum y_i^2\\
\end{pmatrix}
\begin{pmatrix}
a\\
b\\
c\\
\end{pmatrix}
=
\begin{pmatrix}
\sum z_i\\
\sum x_i z_i\\
\sum y_i z_i\\
\end{pmatrix}
$$

#### Question 2


<font size = 3 color = black face = "Times New Roman">Solve the problem as $Ax = b$

In [10]:
# Define the data
x = np.array([0,0,1,2,2,2])
y = np.array([0,1,0,0,1,2])
z = np.array([1.42,1.85,0.78,0.18,0.60,1.05])

In [11]:
# Define matrix A.
A = np.zeros((3,3))
A[0][0] = len(x)
A[1][1] = (x**2).sum()
A[2][2] = (y**2).sum()
A[0][1] = A[1][0] = x.sum()
A[0][2] = A[2][0] = y.sum()
A[1][2] = A[2][1] = (x*y).sum()

In [12]:
# Define matrix b
b_matrix = np.array([z.sum(), (x*z).sum(), (y*z).sum()])

In [13]:
a_matrix = np.dot(np.linalg.inv(A),b_matrix)
print(f'a = {a_matrix[0]}\nb = {a_matrix[0]}\nc = {a_matrix[1]}')

a = 1.4132558139534885
b = 1.4132558139534885
c = -0.6213953488372095


<font size = 3 color = black face = "Times New Roman">Hence we have determined a,b,c.